src/main/sensors/rpm_filter.c

```diff
@@ -0,0 +1,204 @@
+#include "drivers/pwm_output_counts.h"
+#include "sensors/rpm_filter.h"
+#include "common/filter.h"
+#include "sensors/gyro.h"
+#include "build/debug.h"
+#include "flight/mixer.h"
+#include "flight/pid.h"
+
+#define RPM_FILTER_MAXHARMONICS 3
+#define RPM_MOTOR_FILTER_CUTOFF 150
+
+#if defined(USE_RPM_FILTER)
+
+static pt1Filter_t rpmFilters[MAX_SUPPORTED_MOTORS];
+
+typedef struct rpmNotchFilter_s
+{
+    uint8_t harmonics;
+    uint8_t minHz;
+    float   q;
+    float   loopTime;
+
+    biquadFilter_t notch[XYZ_AXIS_COUNT][MAX_SUPPORTED_MOTORS][RPM_FILTER_MAXHARMONICS];
+} rpmNotchFilter_t;
+
+static float   erpmToHz;
+static float   filteredMotorErpm[MAX_SUPPORTED_MOTORS];
+static uint8_t numberFilters;
+static uint8_t numberRpmNotchFilters;
+static uint8_t filterUpdatesPerIteration;
+static float   pidLooptime;
+static rpmNotchFilter_t filters[2];
+static rpmNotchFilter_t* gyroFilter;
+static rpmNotchFilter_t* dtermFilter;
+
+
+PG_REGISTER_WITH_RESET_FN(rpmFilterConfig_t, rpmFilterConfig, PG_RPM_FILTER_CONFIG, 1);
+
+void pgResetFn_rpmFilterConfig(rpmFilterConfig_t *config)
+{
+    config->gyro_rpm_notch_harmonics = 3;
+    config->gyro_rpm_notch_min = 100;
+    config->gyro_rpm_notch_q = 50;
+
+    config->dterm_rpm_notch_harmonics = 1;
+    config->dterm_rpm_notch_min = 100;
+    config->dterm_rpm_notch_q = 50;
+}
+
+static void rpmNotchFilterInit(rpmNotchFilter_t* filter, int harmonics, int minHz, int q, float looptime)
+{
+    filter->harmonics = harmonics;
+    filter->minHz = minHz;
+    filter->q = q / 10.0f;
+    filter->loopTime = looptime;
+
+    for (int axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
+        for (int motor = 0; motor < getMotorCount(); motor++) {
+            for (int i = 0; i < harmonics; i++) {
+                biquadFilterInit(
+                    &filter->notch[axis][motor][i], minHz * i, looptime, filter->q, FILTER_NOTCH);
+            }
+        }
+    }
+}
+
+void rpmFilterInit(const rpmFilterConfig_t *config)
+{
+    pidLooptime = gyro.targetLooptime * pidConfig()->pid_process_denom;
+    numberRpmNotchFilters = 0;
+    if (config->gyro_rpm_notch_harmonics) {
+        gyroFilter = &filters[numberRpmNotchFilters++];
+        rpmNotchFilterInit(gyroFilter, config->gyro_rpm_notch_harmonics,
+                           config->gyro_rpm_notch_min, config->gyro_rpm_notch_q, gyro.targetLooptime);
+    }
+    if (config->dterm_rpm_notch_harmonics) {
+        dtermFilter = &filters[numberRpmNotchFilters++];
+        rpmNotchFilterInit(dtermFilter, config->dterm_rpm_notch_harmonics,
+                           config->dterm_rpm_notch_min, config->dterm_rpm_notch_q, pidLooptime);
+    }
+
+    for (int i = 0; i < getMotorCount(); i++) {
+        pt1FilterInit(&rpmFilters[i], pt1FilterGain(RPM_MOTOR_FILTER_CUTOFF, gyro.targetLooptime));
+    }
+
+    const float secondsPerMinute = 60.0f;
+    const float erpmPerLsb = 100.0f;
+    erpmToHz = erpmPerLsb / secondsPerMinute  / (motorConfig()->motorPoleCount / 2.0f);
```

Comment:
Probably a good idea to document somewhere in the instructions for this that `motorPoleCount` is used - otherwise users with non-default motor pole counts will be caught out.

Reply:
Yeah I agree. I will need to write up a wiki doc on dshot and rpm filtering.